### PR TITLE
MRG: Improve decoding doc and sphinx process

### PR DESCRIPTION
**Pull request by larsoner**
This PR:

1. Converts `manual/decoding.rst` into a tutorial by folding it and `examples/decoding/plot_decoding_csp_space.py` into `tutorials/plot_sensors_decoding.py`
2. Makes our doc building strict about warnings so we stop producing broken links that need to be fixed at release time
3. Removes an image from the channel interpolation doc that did not add much understanding (and in partial builds produces a warning)
4. Fixes the `@deprecated` decorator to properly set the number of spaces to avoid a build warning
5. Improves `conf.py` to turn warnings into errors during example runs (we should not have un-caught warnings in examples)
6. Various minor fixes for doc building to avoid warnings

### CODE REVIEW

**Comment by larsoner** [comment]
https://10182-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_sensors_decoding.html

**Comment by massich** [comment]
I build it locally with `make html_dev-noplot` and I got these warnings:

```
/home/sik/code/mne-python/doc/auto_tutorials/plot_morph_stc.rst:115: WARNING: image file not readable: auto_tutorials/../../_images/sphx_glr_plot_morph_volume_stc_001.png
/home/sik/code/mne-python/doc/auto_tutorials/plot_morph_stc.rst:251: WARNING: image file not readable: auto_tutorials/../../_images/sphx_glr_plot_morph_surface_stc_001.png
/home/sik/code/mne-python/mne/source_estimate.py:docstring of mne.SourceEstimate.morph:10: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/sik/code/mne-python/mne/source_estimate.py:docstring of mne.SourceEstimate.morph_precomputed:10: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/sik/code/mne-python/mne/source_estimate.py:docstring of mne.VectorSourceEstimate.morph:10: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/sik/code/mne-python/mne/source_estimate.py:docstring of mne.VectorSourceEstimate.morph_precomputed:10: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/sik/code/mne-python/doc/manual/channel_interpolation.rst:: WARNING: image file not readable: manual/../../_images/sphx_glr_plot_interpolate_bad_channels_002.png
/home/sik/code/mne-python/doc/manual/decoding.rst:: WARNING: image file not readable: manual/../../_images/sphx_glr_plot_linear_model_patterns_001.png
/home/sik/code/mne-python/doc/manual/decoding.rst:: WARNING: image file not readable: manual/../../_images/sphx_glr_plot_linear_model_patterns_002.png
/home/sik/code/mne-python/doc/manual/decoding.rst:: WARNING: image file not readable: manual/../../_images/sphx_glr_plot_sensors_decoding_001.png
/home/sik/code/mne-python/doc/manual/decoding.rst:: WARNING: image file not readable: manual/../../_images/sphx_glr_plot_sensors_decoding_004.png
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] whats_new                                                                                                                     
/home/sik/code/mne-python/doc/whats_new.rst:108: WARNING: py:func reference target not found: mne.beamformer.make_lcmv`
```
But circle passes without issue.

**Comment by larsoner** [comment]
I don't see it locally either, maybe you need a newer numpydoc or sphinx

**Comment by massich** [comment]
Actually, I mess up. It was my fault.

**Comment by massich** [comment]
But I'm trying to compile it now, and it fails with this packages

sphinx                    1.7.6                    py36_0  
sphinx-bootstrap-theme    0.6.5                     <pip>
sphinx-fontawesome        0.0.6                     <pip>
sphinx-gallery            0.2.0                     <pip>
sphinxcontrib             1.0                      py36_1  
sphinxcontrib-websupport  1.1.0                    py36_1

**Comment by massich** [comment]
This is my log:
```
~/code/mne-python/doc doc* 15s
(mne) ❯ make html_dev-noplot                                                                          
BUILD_DEV_HTML=1 sphinx-build -D plot_gallery=0 -b html -d _build/doctrees  -n -W -T --keep-going . _build/html
Running Sphinx v1.8.1
loading pickled environment... failed: build environment version not current
[autosummary] generating autosummary for: advanced_setup.rst, auto_examples/connectivity/plot_cwt_sensor_connectivity.rst, auto_examples/connectivity/plot_mixed_source_space_connectivity.rst, auto_examples/connectivity/plot_mne_inverse_coherence_epochs.rst, auto_examples/connectivity/plot_mne_inverse_connectivity_spectrum.rst, auto_examples/connectivity/plot_mne_inverse_label_connectivity.rst, auto_examples/connectivity/plot_mne_inverse_psi_visual.rst, auto_examples/connectivity/plot_sensor_connectivity.rst, auto_examples/datasets/plot_brainstorm_data.rst, auto_examples/datasets/plot_hf_sef_data.rst, ..., manual/time_frequency.rst, manual/visualization.rst, martinos.rst, mne_cpp.rst, python_reference.rst, references.rst, tutorials/command_line.rst, tutorials/philosophy.rst, tutorials/report.rst, whats_new.rst
loading intersphinx inventory from https://docs.python.org/3/objects.inv...
loading intersphinx inventory from https://www.numpy.org/devdocs/objects.inv...
loading intersphinx inventory from https://scipy.github.io/devdocs/objects.inv...
loading intersphinx inventory from https://matplotlib.org/objects.inv...
loading intersphinx inventory from https://scikit-learn.org/stable/objects.inv...
loading intersphinx inventory from http://docs.enthought.com/mayavi/mayavi/objects.inv...
loading intersphinx inventory from http://nipy.org/nibabel/objects.inv...
loading intersphinx inventory from http://nilearn.github.io/objects.inv...
loading intersphinx inventory from https://pysurfer.github.io/objects.inv...
loading intersphinx inventory from https://pandas.pydata.org/pandas-docs/stable/objects.inv...
loading intersphinx inventory from http://www.statsmodels.org/stable/objects.inv...
loading intersphinx inventory from http://nipy.org/dipy/objects.inv...
generating gallery...
generating gallery for auto_examples/connectivity... [100%] plot_mixed_source_space_connectivity.py                                      
generating gallery for auto_examples/datasets... [100%] plot_opm_rest_data.py                                                            
generating gallery for auto_examples/decoding... [100%] plot_receptive_field_mtrf.py                                                     
generating gallery for auto_examples/forward... [100%] plot_left_cerebellum_volume_source.py                                             
generating gallery for auto_examples/inverse... [100%] plot_tf_lcmv.py                                                                   
generating gallery for auto_examples/io... [100%] plot_objects_from_arrays.py                                                            
generating gallery for auto_examples/preprocessing... [100%] plot_define_target_events.py                                                
generating gallery for auto_examples/realtime... [100%] rt_feedback_server.py                                                            
generating gallery for auto_examples/simulation... [100%] plot_simulate_raw_data.py                                                      
generating gallery for auto_examples/stats... [100%] plot_cluster_stats_evoked.py                                                        
generating gallery for auto_examples/time_frequency... [100%] plot_time_frequency_simulated.py                                           
generating gallery for auto_examples/visualization... [100%] plot_3d_to_2d.py                                                            
generating gallery for auto_tutorials... [100%] plot_background_filtering.py                                                             
generating MNE command help ... [100%] mne_browse_raw.py                                                                                 
[Done]
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 636 source files that are out of date
updating environment: 636 added, 0 changed, 0 removed
reading sources... [100%] whats_new                                                                                                      
looking for now-outdated files... none found
pickling environment... done
checking consistency... /home/sik/code/mne-python/doc/auto_examples/decoding/plot_decoding_csp_space.rst: WARNING: document isn't included in any toctree
done
preparing documents... done
/home/sik/miniconda3/envs/mne/lib/python3.6/site-packages/sphinx/builders/html.py:130: RemovedInSphinx20Warning: builder.css_files is deprecated. Please use app.add_stylesheet() instead.
  RemovedInSphinx20Warning)
writing output... [100%] whats_new                                                                                                       
generating indices... genindex py-modindex
writing additional pages... search
copying images... [100%] tutorials/mne-report.png                                                                                        
copying downloadable files... [100%] auto_tutorials/plot_whitened.ipynb                                                                  
copying static files... done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build finished with problems, 1 warning.
embedding documentation hyperlinks...
embedding documentation hyperlinks for auto_examples... [100%] plot_sensor_regression.html                                               
embedding documentation hyperlinks for auto_tutorials... [100%] plot_brainstorm_auditory.html                                            
Makefile:60: recipe for target 'html_dev-noplot' failed
make: *** [html_dev-noplot] Error 1
```

**Comment by larsoner** [comment]
`make clean` then try again

**Comment by massich** [comment]
yes! my bad. I had a messy tree.

**Comment by larsoner** [comment]
Yes the `warning` about a file that does not exist in this branch was the clue :)

**Comment by mmagnuski** [comment]
The tutorial looks great!
One tiny nitpick is that this warning box color makes it hard to read the object name:
image
moreover, if we agree on the warning content it should be also mentioned in `mne.decoding.FilterEstimator` docs too. Last, if one knows what they are doing and is careful about epoch length do we still discourage the usage of `FilterEstimator`?

File: tutorials/plot_morph_stc.py [hunk 1]
```diff
@@ -100,12 +100,6 @@
 
     - Save a SourceMorph object to disk
 
-Here is an example result for morphing **volume** source estimates:
-
-:ref:`sphx_glr_auto_examples_inverse_plot_morph_volume_stc.py`
-
-.. image:: ../../_images/sphx_glr_plot_morph_volume_stc_001.png
```
Comment (agramfort) [inline tutorials/plot_morph_stc.py]: why did you remove this? and same remark just below?

File: tutorials/plot_sensors_decoding.py [hunk 2]
```diff
@@ -50,20 +60,243 @@
...
+# Mean scores across cross-validation splits
+score = np.mean(scores, axis=0)
+print('Spatio-temporal: %0.1f%%' % (100 * score,))
+
+###############################################################################
+# PSDEstimator
+# ^^^^^^^^^^^^
+# The :class:`mne.decoding.PSDEstimator`
+# computes the power spectral density (PSD) using the multitaper
+# method. It takes a 3D array as input, it into 2D and computes the PSD.
```
Comment (agramfort) [inline tutorials/plot_sensors_decoding.py]: it into?

File: tutorials/plot_sensors_decoding.py [hunk 3]
```diff
@@ -50,20 +60,243 @@
...
+#
+# .. topic:: Examples
+#
+#     * :ref:`sphx_glr_auto_examples_decoding_plot_decoding_spoc_CMC.py`
+#
+# xDAWN
+# ^^^^^
+# :class:`mne.preprocessing.Xdawn` is a spatial filtering method designed to
+# improve the signal to
+# signal + noise ratio (SSNR) of the ERP responses. Xdawn was originally
```
Comment (agramfort) [inline tutorials/plot_sensors_decoding.py]: can you add the ref to paper like fro SPoC?

File: tutorials/plot_sensors_decoding.py [hunk 4]
```diff
@@ -50,20 +60,243 @@
...
+# Decoding over time
+# ==================
+#
+# This strategy consists in fitting a multivariate predictive model on each
+# time instant and evaluating its performance at the same instant on new
+# epochs. The :class:`mne.decoding.SlidingEstimator` will take as input a
+# pair of features :math:`X` and targets :math:`y`, where :math:`X` has
+# more than 2 dimensions. For decoding over time the data :math:`X`
+# is the epochs data of shape n_epochs x n_channels x n_times. As the
+# last dimension of :math:`X` is the time an estimator will be fit
```
Comment (agramfort) [inline tutorials/plot_sensors_decoding.py]: the time
->
 the time,

File: tutorials/plot_sensors_decoding.py [hunk 5]
```diff
@@ -99,15 +331,28 @@
...
 #
-# The idea is to fit the models on each time instant and see how it
-# generalizes to any other time point.
+# The object to for Temporal generalization is
+# :class:`mne.decoding.GeneralizingEstimator`. It expects as input :math:`X`
+# and :math:`y` (similarly to :class:`~mne.decoding.SlidingEstimator`) but
+# generates predictions from each model for all time instants. The class
+# :class:`~mne.decoding.GeneralizingEstimator` is generic and will treat the
+# last dimension as the one to be used for generalization testing. For
+# convenience, here, we refer to it different tasks. If :math:`X` corresponds
```
Comment (agramfort) [inline tutorials/plot_sensors_decoding.py]: we refer to it different tasks
->
we refer to it as different tasks

File: tutorials/plot_sensors_decoding.py [hunk 6]
```diff
@@ -1,12 +1,29 @@
...
+Design philosophy
+=================
+Decoding (a.k.a. MVPA) in MNE largely follows the machine
+learning API of the scikit-learn package.
+Each estimator implements ``fit``, ``transform``, ``fit_transform``, and
+(optionally) ``inverse_transform`` methods. For more details on this design,
+visit scikit-learn_.
+
+For ease of comprehension, we will denote instantiations of the class using
+the same name as the class but in small caps instead of camel cases.
```
Comment (agramfort) [inline tutorials/plot_sensors_decoding.py]: Can you insert a ref here to our preprint:

https://hal.archives-ouvertes.fr/hal-01848442/

@kingjr what would be a full adequate ref?

File: tutorials/plot_morph_stc.py (hunk 1 above)
Comment (larsoner) [reply tutorials/plot_morph_stc.py]: It does not add much understanding, people should just go to the proper tutorial.

And from a practical standpoint, it emits a sphinx warning any time doing a partial build, so it makes it so we can't get the benefit of turning warnings into errors in general.

**Comment by agramfort** [comment]
ok then

File: tutorials/plot_sensors_decoding.py (hunk 6 above)
Comment (larsoner) [reply tutorials/plot_sensors_decoding.py]: I'll just put "in press" for now

**Comment by larsoner** [comment]
> if we agree on the warning content it should be also mentioned in mne.decoding.FilterEstimator docs too. Last, if one knows what they are doing and is careful about epoch length do we still discourage the usage of FilterEstimator?

Yes, so I've moved the warning to the `FilterEstimator` class and made it less stringent

**Comment by larsoner** [comment]
Forthcoming commit will also improve things by turning warnings during example runs into errors, so that our examples in doc building should (become, with these changes, and from here on out) stay clean

File: examples/inverse/plot_morph_volume_stc.py [hunk 7]
```diff
@@ -42,6 +43,8 @@
 
 print(__doc__)
 
+print(len(warnings.filters))
```
Comment (mmagnuski) [inline examples/inverse/plot_morph_volume_stc.py]: this is temporary, right?

File: examples/inverse/plot_morph_volume_stc.py [hunk 8]
```diff
@@ -158,3 +161,4 @@
                                          subjects_dir=subjects_dir).apply(stc)
 
 plt.show()
+print(len(warnings.filters))
```
Comment (mmagnuski) [inline examples/inverse/plot_morph_volume_stc.py]: same here

**Comment by larsoner** [comment]
Okay CIs are finally green and I think I've addressed all comments. Okay to merge @agramfort ?

**Comment by larsoner** [comment]
I'll actually go ahead and merge and we can tweak docs further if we need to for release (hopefully in the next couple of days!)

File: examples/inverse/plot_mixed_source_space_inverse.py [hunk 9]
```diff
@@ -52,22 +49,19 @@
...
-# But we already have one saved:
-
-src = mne.read_source_spaces(op.join(bem_dir, 'sample-oct-6-src.fif'))
+# Get a surface-based source space, here with few source points for speed:
+src = mne.setup_source_space(subject, spacing='oct5',
+                             add_dist=False, subjects_dir=subjects_dir)
 
 # Now we create a mixed src space by adding the volume regions specified in the
 # list labels_vol. First, read the aseg file and the source space bounds
 # using the inner skull surface (here using 10mm spacing to save time):
```
Comment (agramfort) [inline examples/inverse/plot_mixed_source_space_inverse.py]: sorry to be a bit later here but I am afraid people with just copy paste these defaults which are here due to running time. Can you make more explicit that `oct5` and `pos=10` are not recommended for real analysis? thx

**Comment by massich** [comment]
We were discussing with @agramfort in the blackboard and we thought that the best approach was:

#### 0.17:
  - brainvison, eeglab: add a parameter `stim_channel=True` + deprecation warning suggesting to use `False`
  - edf: `stim_channel='default'` + deprecation warning suggesting to use `None`

#### 0.18:
  - brainvison, eeglab: `stim_channel=False` + deprecation warning saying that the parmeter would disspear
  - edf: if `stim_channel` is differnt than `None` warn that the behaviour would disapear

#### 0.19:
  - Remove synthesis


---
What you propose as a different timeline?

**Comment by massich** [comment]
I mean, if we agree that our users wont mind I'm fine to remove the whole thing in 0.18. The question is how?

**Comment by larsoner** [comment]
(Wrong PR to comment on BTW) to be faster I would do:

0.17:

- Add parameter to bv/eeglab: default None, say accepted values in docstring are True | False
- deprecation warning for all readers that synth will be disabled in 0.18

0.18:

- bv/eeglab, the only accepted value in docstring is `False` (but our default is None), if they pass False warn them to stop passing the parameter
- remove synth for all readers, remove param for edf

0.19:

- remove param for bv/eeglab
